### PR TITLE
Allocate BN Buf with Rsa Verification

### PR DESCRIPTION
**Pull request by lsubash**
Currently BN buf is statically defined and this
would increase stack sizes. Allocate the required memory.

Signed-off-by: Subash Lakkimsetti <subash.lakkimsetti@intel.com>

### CODE REVIEW

File: BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c [hunk 1]
```diff
@@ -54,6 +56,10 @@ int VerifyRsaSignature (CONST PUB_KEY_HDR *PubKeyHdr, CONST SIGNATURE_HDR *Signa
     return err;
   }
 
+  // Allocate BN Buf of size 6K bytes
+  bn_buf_sz = 6*1024;
+  bn_buf = AllocatePool (bn_buf_sz);
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c]: Please check NULL return. No need to do ZeroMem?

File: BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c [hunk 2]
```diff
@@ -63,37 +69,44 @@ int VerifyRsaSignature (CONST PUB_KEY_HDR *PubKeyHdr, CONST SIGNATURE_HDR *Signa
   bn_rsa_e = (IppsBigNumState *)(bn_buf + sz_adj);
   sz_adj = sz_adj + ((sz_e + sizeof(Ipp32u)) & ~(sizeof(Ipp32u)-1));
 
-  if (sz_adj > sizeof(bn_buf)) {
+  if (sz_adj > bn_buf_sz) {
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c]: Why don't you calculate sz_adj first, then allocate necessary memory?

File: BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c (hunk 1 above)
Comment (mauricema) [reply BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c]: also there is API available to get the required buffer size in IPP.
Please try to use API to get it and then allocate it.

File: BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c (hunk 2 above)
Comment (mauricema) [reply BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c]: let us try to allocate the required size from beginning.

**Comment by lsubash** [comment]
Thanks Maurice and Aiden. I would check for IPP API's  to get required size.

**Comment by lsubash** [comment]
> Pls use ippsRSA_GetBufferSizePublicKey() to get the buffer size.

Thanks Maurice. I have done scratch buf allocation using this API.

**Comment by myguitar** [comment]
One concern. FreeMem does not de-allocate memory. Multiple MemAlloc (somewhat big here) may cause an issue before memory init.
@mauricema Isn't AllocateTemporaryMemory() better option?

**Comment by lsubash** [comment]
> One concern. FreeMem does not de-allocate memory. Multiple MemAlloc (somewhat big here) may cause an issue before memory init.
> @mauricema Isn't AllocateTemporaryMemory() better option?

This API can be used. We should follow an order while deallocation using FreeTemporaryMemory.

File: BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c [hunk 3]
```diff
@@ -54,49 +56,71 @@ int VerifyRsaSignature (CONST PUB_KEY_HDR *PubKeyHdr, CONST SIGNATURE_HDR *Signa
...
+  bn_buf_ptr   = bn_buf;
+  rsa_key_s    = (IppsRSAPublicKeyState*) bn_buf_ptr;
+  bn_buf_ptr   = bn_buf_ptr + sz_rsa;
+  bn_rsa_n     = (IppsBigNumState *) bn_buf_ptr;
+  bn_buf_ptr   = bn_buf_ptr + sz_n;
+  bn_rsa_e     = (IppsBigNumState *) bn_buf_ptr;
+
   err = ippsBigNumInit(mod_len / sizeof(Ipp32u), bn_rsa_n);
   if (err != ippStsNoErr) {
+    FreeTemporaryMemory (bn_buf);
```
Comment (mauricema) [inline BootloaderCommonPkg/Library/IppCryptoLib/rsa_verify.c]: can we have a better flow to handle error ?  too many duplicated free memory here.

**Comment by mauricema** [comment]
> One concern. FreeMem does not de-allocate memory. Multiple MemAlloc (somewhat big here) may cause an issue before memory init.
> @mauricema Isn't AllocateTemporaryMemory() better option?

Yes, we should use AllocateTemporaryMemory().

**Comment by lsubash** [comment]
Updated error handling with this PR.

**Pull request merged**